hail/src/main/scala/is/hail/types/physical/stypes/concrete/SInsertFieldsStruct.scala

```diff
@@ -0,0 +1,159 @@
+        case Some(idx) =>
+          val et = insertedFields(idx)._2
+          renamedInsertedFields(idx) = ((newName, et.copy(st = et.st.castRename(newField.typ))))
+        case None => parentPassThroughFieldBuilder += ((oldName, (newName, newField.typ)))
+      }
+    }
+
+    val parentPassThroughMap = parentPassThroughFieldBuilder.result().toMap
+    val parentCastType = TStruct(parentType.fieldNames.map(f => parentPassThroughMap.getOrElse(f, (f, parentType.fieldType(f)))): _*)
+    val renamedParentType = parent.castRename(parentCastType)
+    SInsertFieldsStruct(ts,
+      renamedParentType.asInstanceOf[SBaseStruct],
+      renamedInsertedFields
+    )
+  }
+}
+
+class SInsertFieldsStructSettable(val st: SInsertFieldsStruct, parent: SStructSettable, newFields: IndexedSeq[EmitSettable]) extends SStructSettable {
+  def get: SInsertFieldsStructCode = new SInsertFieldsStructCode(st, parent.load().asBaseStruct, newFields.map(_.load))
+
+  def settableTuple(): IndexedSeq[Settable[_]] = parent.settableTuple() ++ newFields.flatMap(_.settableTuple())
+
+  def loadField(cb: EmitCodeBuilder, fieldIdx: Int): IEmitCode = {
+    st.getFieldIndexInNewOrParent(fieldIdx) match {
+      case Left(parentIdx) => parent.loadField(cb, parentIdx)
+      case Right(newFieldsIdx) => newFields(newFieldsIdx).toI(cb)
+    }
+  }
+
+  def isFieldMissing(fieldIdx: Int): Code[Boolean] =
+    st.getFieldIndexInNewOrParent(fieldIdx) match {
+      case Left(parentIdx) => parent.isFieldMissing(parentIdx)
+      case Right(newFieldsIdx) => newFields(newFieldsIdx).m
+    }
+
+  def store(cb: EmitCodeBuilder, pv: SCode): Unit = {
+    val sifc = pv.asInstanceOf[SInsertFieldsStructCode]
+    parent.store(cb, sifc.parent)
+    newFields.zip(sifc.newFields).foreach { case (settable, code) => cb.assign(settable, code) }
+  }
+}
+
+class SInsertFieldsStructCode(val st: SInsertFieldsStruct, val parent: SBaseStructCode, val newFields: IndexedSeq[EmitCode]) extends SBaseStructCode {
+  override def makeCodeTuple(cb: EmitCodeBuilder): IndexedSeq[Code[_]] = parent.makeCodeTuple(cb) ++ newFields.flatMap(_.makeCodeTuple(cb))
+
+  override def memoize(cb: EmitCodeBuilder, name: String): SInsertFieldsStructSettable = {
+    new SInsertFieldsStructSettable(st, parent.memoize(cb, name + "_parent").asInstanceOf[SStructSettable], newFields.indices.map { i =>
+      val code = newFields(i)
+      val es = cb.emb.newEmitLocal(s"${ name }_nf_$i", code.emitType)
+      es.store(cb, code)
+      es
+    })
+  }
+
+  override def memoizeField(cb: EmitCodeBuilder, name: String): SInsertFieldsStructSettable = {
+    new SInsertFieldsStructSettable(st, parent.memoizeField(cb, name + "_parent").asInstanceOf[SStructSettable], newFields.indices.map { i =>
+      val code = newFields(i)
+      val es = cb.emb.newEmitField(s"${ name }_nf_$i", code.emitType)
+      es.store(cb, code)
+      es
+    })
+  }
+
+  override def insert(newType: TStruct, fields: (String, EmitCode)*): SBaseStructCode = {
```

Comment:
I think you're right that this could handle the stack struct case (though the empty/unreachable parent seems a little hacky). Seems a little cleaner to have the separate classes, though, even with a bit of overlap.